packages/aws-cdk/lib/archive.ts

```diff
@@ -1,23 +1,45 @@
 import archiver = require('archiver');
 import crypto = require('crypto');
 import fs = require('fs-extra');
+import glob = require('glob');
+import path = require('path');
 
 export function zipDirectory(directory: string, outputFile: string): Promise<void> {
   return new Promise((ok, fail) => {
     const output = fs.createWriteStream(outputFile);
     const archive = archiver('zip');
+
     // The below options are needed to support following symlinks when building zip files:
-    // -  nodir: This will prevent symlinks themselves from being copied into the zip.
+    // - nodir: This will prevent symlinks themselves from being copied into the zip.
     // - follow: This will follow symlinks and copy the files within.
     const globOptions = {
       dot: true,
       nodir: true,
       follow: true,
-      cwd: directory
+      cwd: directory,
     };
-    archive.glob('**', globOptions);
-    archive.pipe(output);
-    archive.finalize();
+    const files = glob.sync('**', globOptions); // The output here is already sorted
+
+    output.on('open', async () => {
+      archive.pipe(output);
+
+      const contents = await Promise.all(files.map(async (file) => {
```

Comment:
This will load all files into memory concurrently. Not sure I like that for very big zips (tens to hundreds of megabytes).

Also, don't see the use of doing this in parallel since this task should be I/O bound, as far as I can estimate.

Can you test and verify whether the speed makes a difference? If it does, then I would prefer some kind of batching (chop the list into batches of 10 and do those in parallel) to keep memory consumption under control.

Reply:
Understand, will have a look at it